### PR TITLE
Added support for mapping selectors

### PR DESCRIPTION
**Pull request by PSpSynedra**
Adds the `composeSelectors` API-Function to create mapping selectors.

### CODE REVIEW

File: test/test_selector.js [hunk 1]
```diff
@@ -189,4 +189,40 @@ suite('selector', () => {
...
+        assert.notEqual(selector({a: 2,b: 2}), firstResult);
+        assert.deepEqual(selector({a: 2,b: 2}), {x: 2,y: 2});
+    });
+    test("composing selectors with custom selector creator", function() {
+        let customEqualityFunctionCalls = 0;
+        const customSelectorCreator = createSelectorCreator(
+            defaultMemoize,
+            (a,b) => {
+                ++customEqualityFunctionCalls;
+                return JSON.stringify(a) == JSON.stringify(b)
```
Comment (ellbee) [inline test/test_selector.js]: I can't figure out the reason behind counting the calls to the equality check function. Also, JSON.stringify(a) == JSON.stringify(b) can be replaced by a === b and the tests still pass.

File: test/test_selector.js [hunk 2]
```diff
@@ -189,4 +189,40 @@ suite('selector', () => {
...
+              y: state => state.b
+        }, customSelectorCreator );
+
+        let firstResult = selector({a: 1,b: 2});
+        assert.deepEqual( firstResult, {x: 1, y: 2});
+        assert.strictEqual(selector({a: 1,b: 2}), firstResult);
+        assert.notStrictEqual(selector({a: 2,b: 2}),firstResult);
+        assert.notEqual(selector({a: 2,b: 2}), firstResult);
+        assert.deepEqual(selector({a: 2,b: 2}), {x: 2,y: 2});
+
```
Comment (ellbee) [inline test/test_selector.js]: Do we need the notEqual and notStrictEqual checks on lines 222 and 223 if we have the deepEqual check on line 224?

**Comment by ellbee** [comment]
Thanks @PSpSynedra, I have been trying this out and it is working great so far. I've just made a couple of line notes about the tests.

File: test/test_selector.js (hunk 2 above)
Comment (PSpSynedra) [inline test/test_selector.js]: That's right, these two checks are not necessary as you pointed out correctly. I will simply remove them.

File: test/test_selector.js (hunk 1 above)
Comment (PSpSynedra) [inline test/test_selector.js]: The stringify is simply to demonstrate that any custom equality function may be used.
The counting only ensures, that the equality function is not being called to often while checking the arguments for equality. But with the other tests still passing, this is no necessity either.
I am gonna simplify the case.

**Comment by ellbee** [comment]
I wonder if `combineSelectors` is a better name than `composeSelectors` for this? It seems to me we are combining the selectors into a single selector rather than composing the selectors together.

**Comment by PSpSynedra** [comment]
Yeah the decision was between the two of them and as we called it combineSelectors internally, well ;-)
But for me combineSelectors is also fine as it actually combines them into a map.
Although the map needs to be provided by the caller already so the actual combining needs to be done by the caller.

Hm *thinking*. 

Maybe createMappingSelector or createStructuredSelector would describe its functionality even better?

**Comment by PSpSynedra** [comment]
No we called it composeSelectors of course, sorry.

**Comment by ellbee** [comment]
Yes, I guess `combineSelectors` might not the right name either!

**Comment by PSpSynedra** [comment]
What about createMappingSelector or createStructuredSelector as mentioned earlier?

**Comment by ellbee** [comment]
Yeah, I agree that they more accurately describe what is going on so I think they are candidates. I have another possibility to throw out, how about `createKeyedSelector`?

**Comment by PSpSynedra** [comment]
I'd rather prefer `createStructuredSelector`, as it allows you to create an arbitrarily structured selector like in:
```javascript
let selector = createStructuredSelector({
   subA: createStructuredSelector({
       selectorA: whatever
       selectorB: whatever
   }),
   subB: createStructuredSelector({
       selectorC: whatever
       selectorD: whatever
   })
});

let values = selector( state );
/* Expect values to be structured like:
   {
       subA: {
           selectorA: value
           selectorB: value
       },
       subB: {
           selectorC: value
           selectorD: value
       }
   }
*/
```

**Comment by ellbee** [comment]
OK, let's go with that then. I'll merge this in later today.

**Comment by ellbee** [comment]
@PSpSynedra, thanks a lot for working on this! It has been merged and released as v1.10.

I'm going to close the PR because I merged the commits locally so I could squash them.

**Comment by PSpSynedra** [comment]
Cool, thanks

**Pull request closed without merging**